sdk/cosmos/azure-cosmos/azure/cosmos/_service_request_retry_policy.py

```diff
@@ -43,28 +45,85 @@ def ShouldRetry(self):
+                # reset the in region retry count
+                self.in_region_retry_count = 0
+                self.failover_retry_count += 1
+                if self.failover_retry_count >= self.total_retries:
+                    return False
 
-            # set location-based routing directive based on retry count
-            # ensuring usePreferredLocations is set to True for retry
-            self.request.route_to_location_with_preferred_location_flag(self.failover_retry_count, True)
+            self.request.last_routed_location_endpoint_within_region = self.request.location_endpoint_to_route
+            if _OperationType.IsReadOnlyOperation(self.request.operation_type):
+                # We just directly got to the next location in case of read requests
+                # We don't retry again on the same region for regional endpoint
+                self.failover_retry_count += 1
+                if self.failover_retry_count >= self.total_retries:
+                    return False
+                # # Check if it is safe to failover to another region
+                self.location_endpoint = self.resolve_next_region_service_endpoint()
+            else:
+                self.location_endpoint = self.resolve_current_region_service_endpoint()
+                # This is the case where both current and previous point to the same writable endpoint
+                # In this case we don't want to retry again, rather failover to the next region
+                if self.request.last_routed_location_endpoint_within_region == self.location_endpoint:
+                    # Since both the endpoints (current and previous) are same, we mark them unavailable
+                    # and refresh the cache.
+                    print("In service request - Both current and previous are same, marking them unavailable")
+                    self.mark_endpoint_unavailable(True)
+                    # Although we are not retrying again in this region
+                    # but since this is the same endpoint, our in region retries are done for this region
+                    # and we reset the in region retry count
+                    self.in_region_retry_count = 0
+                    self.failover_retry_count += 1
+                    # # Check if it is safe to failover to another region
+                    if self.failover_retry_count >= self.total_retries:
+                        return False
+                    self.location_endpoint = self.resolve_next_region_service_endpoint()
 
-            # Resolve the endpoint for the request and pin the resolution to the resolved endpoint
-            # This enables marking the endpoint unavailability on endpoint failover/unreachability
-            self.location_endpoint = self.global_endpoint_manager.resolve_service_endpoint(self.request)
+            print("In service request - Location endpoint: {}".format(self.location_endpoint))
```

Comment:
And another

Reply:
Actually there's lots - might need to do a find-all

Reply:
Will remove all of these later on with a find-all :)